src/Processors/Transforms/FillingTransform.cpp

```diff
@@ -207,44 +227,94 @@ void FillingTransform::transform(Chunk & chunk)
         {
             auto old_column = old_columns[pos]->convertToFullColumnIfConst();
             new_columns.push_back(old_column);
+            res_map[pos] = {&new_mutable_columns, new_mutable_columns.size()};
             new_mutable_columns.push_back(old_column->cloneEmpty()->assumeMutable());
         }
     };
 
+    Block interpolate_block;
+
+    auto interpolate = [&]()
+    {
+        if (interpolate_description)
+        {
+            interpolate_block.clear();
+
+            if (interpolate_description->input_positions.size())
+            {
+                for (const auto & [col_pos, name_type] : interpolate_description->input_positions)
+                {
+                    MutableColumnPtr column = name_type.type->createColumn();
+                    auto [res_columns, pos] = res_map[col_pos];
+                    size_t size = (*res_columns)[pos]->size();
+                    if (size == 0)
+                    {
+                        if (last_row.size() > col_pos && last_row[col_pos]->size())
+                            column->insertFrom(*last_row[col_pos], 0);
+                        else
+                            column->insertDefault();
+                    }
+                    else
+                        column->insertFrom(*(*res_columns)[pos], size - 1);
+
+                    interpolate_block.insert({std::move(column), name_type.type, name_type.name});
+                }
+            }
+            else /// all INTERPOLATE expressions are constants
+            {
+                /// dirty hack - we need at least one column with one row to execute actions on block
+                DataTypePtr dt = std::make_shared<DataTypeUInt64>();
+                interpolate_block.insert({dt->createColumnConst(1, dt->getDefault()), dt, "dummy"});
+            }
+
+            interpolate_description->actions->execute(interpolate_block);
+        }
+    };
+
     if (generate_suffix)
     {
         const auto & empty_columns = input.getHeader().getColumns();
         init_columns_by_positions(empty_columns, old_fill_columns, res_fill_columns, fill_column_positions);
+        init_columns_by_positions(empty_columns, old_interpolate_columns, res_interpolate_columns, interpolate_column_positions);
         init_columns_by_positions(empty_columns, old_other_columns, res_other_columns, other_column_positions);
 
         if (first)
             filling_row.initFromDefaults();
 
         if (should_insert_first && filling_row < next_row)
-            insertFromFillingRow(res_fill_columns, res_other_columns, filling_row);
+        {
+            interpolate();
+            insertFromFillingRow(res_fill_columns, res_interpolate_columns, res_other_columns, filling_row, interpolate_block);
+        }
 
-        while (filling_row.next(next_row))
-            insertFromFillingRow(res_fill_columns, res_other_columns, filling_row);
+        for (interpolate(); filling_row.next(next_row); interpolate())
```

Comment:
Well, I think it is not very readable to call interpolate() inside the for loop.
Could we, please, do it before/in the loop body?